### ISSUE TITLE
MDS liveness probe configuration

### ISSUE THREAD

**Issue by SerialVelocity** [rook/rook#6355]
**Is this a bug report or feature request?**
* Feature Request

**What should the feature do:**
Add a configuration option to allow configuring/disabling the liveness probe for MDSs

**What is use case behind this feature:**
All other daemons now have configuration options for this, it would be good to be able to disable or change the timeout

Relevant PRs:
#5789 (Mon/OSD/Status liveness probe configuration)
#5836 (ObjectStore liveness probe configuration)

**Environment**:

Bare metal

**Comment by SerialVelocity** [comment]
@leseb After upgrading to Kubernetes 1.20 this has become pretty important. The ceph status command doesn't complete within 1s and the "fix" in Kubernetes for exec probes causes my mds pods to die constantly.

**Comment by SerialVelocity** [comment]
#6883 fixes this, right?

**Comment by travisn** [comment]
Yes, this is expected to be fixed by #6883, which was included in v1.5.5. Let us know if you see any issues with it.

**Comment by SerialVelocity** [comment]
@travisn @jshen28 I don't think that PR works. `v1.Container` is a struct, not an interface, so you have to use the return value of `ConfigureLivenessProbe`:
https://github.com/rook/rook/blob/v1.5.5/pkg/operator/ceph/file/mds/spec.go#L44
```diff
-	config.ConfigureLivenessProbe(cephv1.KeyMds, mdsContainer, c.clusterSpec.HealthCheck)
+	mdsContainer = config.ConfigureLivenessProbe(cephv1.KeyMds, mdsContainer, c.clusterSpec.HealthCheck)
````

**Comment by SerialVelocity** [comment]
Also, not sure if it is related to the liveness probe or not, but I have a tonne of defunct ldconfig processes left over in the ceph-mds container. @travisn Do you mind checking if it happens in your mds container as well?

```sh
> ps axco command | sort | grep -vE '(COMMAND|bash|grep|ps|sort|uniq)' | uniq -c
      1 ceph-mds
    111 ldconfig <defunct>
```

**Comment by travisn** [comment]
@SerialVelocity Good catch, the container must be assigned on the return from the method. @jshen28 Want to open a follow-up to #6883 to fix this?

**Comment by travisn** [comment]
> Also, not sure if it is related to the liveness probe or not, but I have a tonne of defunct ldconfig processes left over in the ceph-mds container. @travisn Do you mind checking if it happens in your mds container as well?
> 
> ```shell
> > ps axco command | sort | grep -vE '(COMMAND|bash|grep|ps|sort|uniq)' | uniq -c
>       1 ceph-mds
>     111 ldconfig <defunct>
> ```

@SerialVelocity I don't see the defunct processes in my mds pods, but it's just a test cluster without any load. You may want to open a ceph tracker issue for that. @joscollin Any idea on this?

**Comment by jshen28** [comment]
I'll definitely look into the bug

**Comment by SerialVelocity** [comment]
Just posting this here so everything is linked:
- Initial attempt at configuring/disabling probes (broken): #6883 (1.5.5 backport: #6942)
- Fix for actually modifying/disabling the probe: #7077 (1.5.6 backport: #7090)

**Comment by joscollin** [comment]
> @SerialVelocity I don't see the defunct processes in my mds pods, but it's just a test cluster without any load. You may want to open a ceph tracker issue for that. @joscollin Any idea on this?

@SerialVelocity 
If you confirm that this is a Ceph related issue (I don't think so), open a tracker with necessary logs and we will check it. Thanks.

**Comment by SerialVelocity** [comment]
@joscollin @travisn It looks like that zombie process issue is because rook is using exec probes. See here:
https://youtu.be/QKI-JRs2RIE?t=1419

**Comment by travisn** [comment]
@SerialVelocity Good call, I'm surprised we haven't had reports of this before. The liveness probes really need to be running with tini so they're cleaned up properly. As a quick short term fix we can at least increase the timeout from the default 1s.

**Comment by leseb** [comment]
Hum wait a second, what is calling `ldconfig`? Our liveness probe is just calling the Ceph CLI like this: `ceph --admin-daemon /run/ceph/ceph-mds.0.asok status`.

I wonder where this `ldconfig` is coming from. I have the impression that we might be mixing things up...

**Comment by travisn** [comment]
> I wonder where this ldconfig is coming from. I have the impression that we might be mixing things up...

Oh right, I overlooked that question in my last response, not sure about that.

**Comment by SerialVelocity** [comment]
Afaik, it is an expected process that runs inside the container. Whether it is by the liveness probe or the main process I am not sure though.
https://securityhub.dev/falco-rules/rook lists it as expected.
